kitsune/search/v2/base.py

```diff
@@ -243,40 +246,68 @@ def prepare_locale(self, instance):
         return ""
 
 
-class SumoSearch(ABC):
-    """Base class for search classes.
-
-    Implements the run() function, which will perform the search.
+class SumoSearchInterface(ABC):
+    """Base interface class for search classes.
 
     Child classes should define values for the various abstract properties this
     class has, relevant to the documents the child class is searching over.
     """
 
-    def __init__(self, **kwargs):
-        self.results_per_page = settings.SEARCH_RESULTS_PER_PAGE
-        self.hits = []
-        self.total = 0
-        self.results = []
-
     @abstractmethod
     def get_index(self):
         """The index or comma-seperated indices to search over."""
-        pass
+        ...
 
     @abstractmethod
     def get_fields(self):
         """An array of fields to search over."""
-        pass
+        ...
 
     @abstractmethod
     def get_highlight_fields_options(self):
         """An array of tuples of fields to highlight and their options."""
-        pass
+        ...
 
     @abstractmethod
     def get_filter(self, **kwargs):
         """A query which filters for all documents to be searched over."""
-        pass
+        ...
+
+    @abstractmethod
+    def build_query(self, **kwargs):
+        """Build a query to search over a specific set of documents."""
+        ...
+
+    @abstractmethod
+    def make_result(self, hit):
+        """Takes a hit and returns a result dictionary."""
+        ...
+
+    @abstractmethod
+    def run(self, **kwargs):
+        """Perform search, placing the results in `self.results`, and the total
+        number of results (across all pages) in `self.total`. Chainable."""
+        ...
+
+
+@dataclass
+class SumoSearch(SumoSearchInterface):
+    """Base class for search classes.
+
+    Implements the run() function, which will perform the search.
+
+    Child classes should define values for the various abstract properties this
+    class inherits, relevant to the documents the child class is searching over.
+    """
+
+    results_per_page: int = settings.SEARCH_RESULTS_PER_PAGE
+    total: int = 0
+    hits: List[Dict] = dfield(default_factory=list)
+    results: List[Dict] = dfield(default_factory=list)
```

Comment:
Since 3.9 it seems to be preferable to use `dict[]` and `list[]` directly: https://docs.python.org/3/library/typing.html?highlight=typing#typing.Dict
> Deprecated since version 3.9: builtins.dict now supports []. See PEP 585 and Generic Alias Type.